docs/application/native/guides/text-input/voice-control-engine.md

```diff
@@ -0,0 +1,527 @@
+
+   ```
+    static int private_data_set_cb(const char* key, const char* data);
+    static int private_data_requested_cb(const char* key, char** data);
+    static int nlu_base_info_requested_cb(const char* key, char** value);
+
+    void
+    int main(int argc, char* argv[])
+    {
+        vce_request_callback_s callback = {
+            /* Add the mandatory callbacks */
+        };
+
+        vce_set_private_data_set_cb(private_data_set_cb);
+        vce_set_private_data_requested_cb(private_data_requested_cb);
+        vce_set_nlu_base_info_requested_cb(nlu_base_info_requested_cb);
+
+        vce_main(argc, argv, &callback);
+    }
+   ```
+
+4. Define the optional callbacks:
+   ```
+    static int
+    private_data_set_cb(const char* key, const char* data)
+    {
+        return 0;	/* private data save succeed */
+    }
+
+    static int
+    private_data_requested_cb(const char* key, char** data)
+    {
+        /* found key */
+        *data = strdup("data");
+        return 0;	/* private data set succeed */
+    }
+
+    static int
+    nlu_base_info_requested_cb(const char* key, char** value)
+    {
+        /* found key */
+        *value = strdup("value");
+        return 0;	/* nlu base info value set succeed */
+    }
+   ```
+
+<a name="parameters"></a>
+## VCE Parameters
+
+You can set the following parameters about the VCE:
+
+- Send the current state using the `vce_send_error()` function and send the error to a engine service user.
+
+  ```
+    void
+    send_error(void)
+    {
+        int ret = VCE_ERROR_NONE;
+
+        ret = vce_send_error(VCE_ERROR_NONE, "msg", NULL);
+        if(VCE_ERROR_NONE != ret)
+            /* Error handling */
+    }
+  ```
+
+- Send the current state using the `vce_set_private_data()` function. Set private data to a engine service user.
+
+  ```
+    void
+    set_private_data(void)
+    {
+        int ret = VCE_ERROR_NONE;
+
+        ret = vce_set_private_data("key", "data");
+        if(VCE_ERROR_NONE != ret)
+            /* Error handling */
+    }
+  ```
+
+
+<a name="info_retrieval"></a>
+## VCE Information Retrieval
+
+You can get the following information about the VCE:
+
+- Get all commands using the `vce_get_foreach_command()` function.
+
+  ```
+    bool
+    command_cb(int id, int type, int format, const char* command, const char* param, int domain, void* user_data)
+    {
+        /* Callback command data */
+        /* Your code */
+    }
+
+    void
+    get_foreach_command(void)
+    {
+        int ret = VCE_ERROR_NONE;
+        int g_vce_command;
+
+        ret = vce_get_foreach_command(g_vce_command, command_cb, NULL);
+        if(VCE_ERROR_NONE != ret)
+            /* Error handling */
+    }
+  ```
+
+- Get command length using the `vce_get_command_count()` function. Get `g_vce_command` data from set_commands()
```

Comment:
Get command length using the `vce_get_command_count()` function. Get `g_vce_command` data from set_commands()
"Please can you add period at the end of the sentence"

Reply:
Done.